pkg/queryfrontend/config.go

```diff
@@ -0,0 +1,149 @@
+// Based on the config type the config is then parsed into a specific cache provider.
+type CacheProviderConfig struct {
+	Type   ResponseCacheProvider `yaml:"type"`
+	Config interface{}           `yaml:"config"`
+}
+
+// NewCacheConfig is a parser that converts a Thanos cache config yaml into a cortex cache config struct.
+func NewCacheConfig(logger log.Logger, confContentYaml []byte) (*queryrange.ResultsCacheConfig, error) {
+	cacheConfig := &CacheProviderConfig{}
+	if err := yaml.UnmarshalStrict(confContentYaml, cacheConfig); err != nil {
+		return nil, errors.Wrap(err, "parsing config YAML file")
+	}
+
+	backendConfig, err := yaml.Marshal(cacheConfig.Config)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal content of cache backend configuration")
+	}
+
+	switch strings.ToUpper(string(cacheConfig.Type)) {
+	case string(INMEMORY):
+		var config InMemoryResponseCacheConfig
+		if err := yaml.Unmarshal(backendConfig, &config); err != nil {
+			return nil, err
+		}
+
+		return &queryrange.ResultsCacheConfig{
+			CacheConfig: cortexcache.Config{
+				EnableFifoCache: true,
+				Fifocache: cortexcache.FifoCacheConfig{
+					MaxSizeBytes: config.MaxSize,
+					MaxSizeItems: config.MaxSizeItems,
+					Validity:     config.Validity,
+				},
+			},
+		}, nil
+	case string(MEMCACHED):
+		var config MemcachedResponseCacheConfig
+		if err := yaml.UnmarshalStrict(backendConfig, &config); err != nil {
+			return nil, err
+		}
+		// TODO(krasi) Add support for it in the cortex module.
+		if config.Memcached.MaxItemSize > 0 {
+			level.Warn(logger).Log("message", "MaxItemSize is not yet supported by the memcached client")
+		}
+
+		return &queryrange.ResultsCacheConfig{
+			CacheConfig: cortexcache.Config{
+				Memcache: cortexcache.MemcachedConfig{
+					Expiration:  config.Expiration,
+					Parallelism: config.Memcached.MaxGetMultiConcurrency,
+					BatchSize:   config.Memcached.MaxGetMultiBatchSize,
+				},
+				MemcacheClient: cortexcache.MemcachedClientConfig{
+					Timeout:        config.Memcached.Timeout,
+					MaxIdleConns:   config.Memcached.MaxIdleConnections,
+					Addresses:      strings.Join(config.Memcached.Addresses, ","),
+					UpdateInterval: config.Memcached.DNSProviderUpdateInterval,
+				},
+				Background: cortexcache.BackgroundConfig{
+					WriteBackBuffer:     config.Memcached.MaxAsyncBufferSize,
+					WriteBackGoroutines: config.Memcached.MaxAsyncConcurrency,
+				},
+			},
+		}, nil
+	default:
+		return nil, errors.Errorf("index cache with type %s is not supported", cacheConfig.Type)
+	}
+}
+
+// Config holds the query frontend configs.
+type Config struct {
+	// PartialResponseStrategy is the default strategy used
+	// when parsing thanos query request.
+	PartialResponseStrategy bool
+
+	CortexFrontendConfig     *cortexfrontend.Config
+	CortexLimits             *cortexvalidation.Limits
+	CortexResultsCacheConfig *queryrange.ResultsCacheConfig
```

Comment:
Not a blocker. Is it necessary to add `Cortex` prefix to these fields?

Reply:
when reading the code later it is a bit more clear